### PR TITLE
SEC-472 Java instanceof: avoid dereferencing null pointer

### PR DESCRIPTION
**Pull request by smowton**
Previously our code was of the form
```
tmp = x->@class_identifier
is_instanceof = x != null && tmp == "A" || tmp == "B" || ...
```

This was harmless as the value read from a null pointer was never used if it was null, but
would present a spurious dereference of a possibly-null pointer, introducing false uncertainty
about the vallue of `tmp`. Therefore we now generate:

```
if x == null:
  is_instanceof = false
else:
  tmp = x->@class_identifier
  is_instanceof = tmp == "A" || tmp == "B" || ...
```

### CODE REVIEW

**Comment by smowton** [comment]
Note this is an alternative solution to https://github.com/diffblue/cbmc/pull/2360, which simply added a null check to both lines, but caused DeepTest's test run to take much longer than usual.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 518ca3ff8a361c174004e762e607ecc5e7385b22).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76800992

**Comment by smowton** [comment]
Looks like this works! Ready for review. TG bump also passing: https://github.com/diffblue/test-gen/pull/1932

File: jbmc/src/java_bytecode/remove_instanceof.cpp [hunk 1]
```diff
@@ -50,17 +51,17 @@ class remove_instanceoft
 /// \param goto_program: program the expression belongs to
 /// \param this_inst: instruction the expression is found at
 /// \return number of instanceof expressions that have been replaced
```
Comment (mgudemann) [inline jbmc/src/java_bytecode/remove_instanceof.cpp]: doxygen needs to be adapted here

File: jbmc/src/java_bytecode/remove_instanceof.cpp [hunk 2]
```diff
@@ -50,17 +51,17 @@ class remove_instanceoft
...
+bool remove_instanceoft::lower_instanceof(
   exprt &expr,
   goto_programt &goto_program,
   goto_programt::targett this_inst)
 {
   if(expr.id()!=ID_java_instanceof)
   {
-    std::size_t replacements=0;
+    bool changed = false;
     Forall_operands(it, expr)
```
Comment (mgudemann) [inline jbmc/src/java_bytecode/remove_instanceof.cpp:66]: maybe chage to ranged for while you're at it

**Review by mgudemann — COMMENTED** [review]
maybe add a unit test that checks that now only a single deref is done

File: jbmc/src/java_bytecode/remove_instanceof.cpp [hunk 3]
```diff
@@ -94,46 +95,93 @@ std::size_t remove_instanceoft::lower_instanceof(
...
-  // Replace the instanceof construct with a conjunction of non-null and the
-  // disjunction of all possible object types. According to the Java
-  // specification, null instanceof T is false for all possible values of T.
+  // According to the Java specification, null instanceof T is false for all
+  // possible values of T.
   // (http://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.20.2)
-  notequal_exprt non_null_expr(
-    check_ptr, null_pointer_exprt(to_pointer_type(check_ptr.type())));
+
+  code_ifthenelset is_null_branch;
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/remove_instanceof.cpp:135]: Not sure `goto_convert`should be used here... Would it be significantly longer to rewrite this as a proper goto program transformation?

File: jbmc/src/java_bytecode/remove_instanceof.cpp [hunk 4]
```diff
@@ -94,46 +95,93 @@ std::size_t remove_instanceoft::lower_instanceof(
...
-  expr = and_exprt(non_null_expr, disjunction(or_ops));
+  else_block.add(
+    code_assignt(instanceof_result_sym.symbol_expr(), disjunction(or_ops)));
+
+  is_null_branch.else_case() = std::move(else_block);
+
+  // Replace the instanceof construct with instanceof_result:
+  expr = instanceof_result_sym.symbol_expr();
+
+  null_message_handlert replaceme;
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/remove_instanceof.cpp]: `remove_instanceoft` whould then probably have a message handler instead of swallowing potential output here (irrelevant if goto_convert wasn't used).

File: jbmc/src/java_bytecode/remove_instanceof.cpp [hunk 5]
```diff
@@ -39,7 +40,7 @@ class remove_instanceoft
   namespacet ns;
   class_hierarchyt class_hierarchy;
 
-  std::size_t lower_instanceof(
+  bool lower_instanceof(
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/remove_instanceof.cpp]: This refactoring should probably be in a separate commit.

Comment (smowton) [reply jbmc/src/java_bytecode/remove_instanceof.cpp]: This isn't refactoring, but rather removing a return value that is no longer needed due to the remaining changes

File: jbmc/src/java_bytecode/remove_instanceof.cpp (hunk 4 above)
Comment (smowton) [reply jbmc/src/java_bytecode/remove_instanceof.cpp]: Since this should never be able to report anything, how about capture the output and `INVARIANT` that nothing was written (and report if it was)?

File: jbmc/src/java_bytecode/remove_instanceof.cpp (hunk 3 above)
Comment (smowton) [reply jbmc/src/java_bytecode/remove_instanceof.cpp:135]: Not much, but it seems worse than building the `if/else` structure I intend then  letting goto-convert build the CFG as it sees fit. See also making stub methods in `jbmc_parse_optionst`, which is already done by making a codet and then `goto_convert`ing it

**Comment by smowton** [comment]
Applied one of Matthias' fixes; replied to @peterschrammel

File: jbmc/src/java_bytecode/remove_instanceof.cpp (hunk 4 above)
Comment (peterschrammel) [reply jbmc/src/java_bytecode/remove_instanceof.cpp]: Shouldn't we then remove the messagehandler from it? We might want to create an issue for this to defer it to a follow-up pr.

Comment (smowton) [reply jbmc/src/java_bytecode/remove_instanceof.cpp]: Well, I imagine for user-supplied code it can spot mistakes they have made and report them, if the frontend hasn't done so first? But here the code is my creation, so any mistakes are a coding mistake rather than something to tell the user about -- hence my intent to use a stringstream and assert that there were no failure reports.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: cf89df7f187290c70ef9ac21d1047b01acc993e4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76834237

**Comment by smowton** [comment]
@peterschrammel applied suggested change to check via an invariant that goto-convert had nothing to complain about.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 37c9da4a21484d14e25a920d3a2a7b0cc5d0f435).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76843088

File: jbmc/src/java_bytecode/remove_instanceof.cpp [hunk 6]
```diff
@@ -94,46 +95,99 @@ std::size_t remove_instanceoft::lower_instanceof(
...
+  // Insert the new test block before it:
+  goto_programt new_check_program;
+  goto_convert(
+    is_null_branch,
+    symbol_table,
+    new_check_program,
+    convert_message_handler,
+    ID_java);
+
+  INVARIANT(
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/remove_instanceof.cpp]: There is actually output into debug() in goto_convert; hence, this will fail with --verbosity 10.
it would be cleaner to simply provide remove_instanceof with a message_handler.

**Comment by smowton** [comment]
The stream-message-handler actually doesn't inherit its verbosity from the command-line -- I've now set it to warning level, so I'm effectively asserting no warnings or errors are generated by the conversion. I quick look at `convert_ifthenelse` in particular and all uses of `warning()` and `error()` in goto_convert*.cpp suggest these really are serious mistakes on my part if they occur.

(Main reason for avoiding adding the message handler -- I'd then have to add one to `remove_exceptions` too, and the resulting messages would be nonsensical to the user -- they would refer to generated code not present in their input, doubly so in the case of usage via `remove_exceptionst`)

**Comment by smowton** [comment]
NB. if #2125 is merged first I will be able to provide a test case for this, showing that the instanceof test no longer produces a spurious `invalid_object` reference! @kroening ... ?

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: b8631b98ee8a46382495ce477092bcecf582c5da).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76921995

**Review by peterschrammel — APPROVED** [review]
Not ideal, but let's go ahead with that.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 8d6b393bb48c27a5d36177d6871b2a803f6d405d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77042416
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 2369df322175810a642a3e44a8777f9008363243).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77152159
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##
